### ISSUE TITLE
Savepoint names can be duplicated

### ISSUE THREAD

**Issue by wizzard0** [go-gorm/gorm#4492]
## Description

https://github.com/go-gorm/gorm/blob/6d64e31965f5c9c383e351683052d7aec68c56b5/finisher_api.go?#L522
In this line, savepoint name is generated from the pointer to the callback passed to the gorm.Transaction function.

If you ever reuse the callback, or the closure happens to be allocated at the same address, the transaction will silently fail.

Please allow to pass the savepoint name, or at least add some randomness there.

**Comment by pitluga** [comment]
This bit me as well. In order to work around it, I kept track of the nested depth of my transactions and uses a different copy of the function. If you do a switch statement, you can get different pointers for different nested depths. Something like:

```
var txfn func(tx *gorm.DB) error

switch depth {
case 0: 
  txfn = func(tx *gorm.DB) error { ... }
case 1: 
  txfn = func(tx *gorm.DB) error { ... }
case n: 
  txfn = func(tx *gorm.DB) error { ... }
default:
  return errors.New("too deep")
}
```

**Comment by phroggyy** [comment]
Re-reported and reproduced in #7173, fix in #7174
